src/core/server/plugins/plugins_system.ts

```diff
@@ -159,7 +159,15 @@ export class PluginsSystem {
       const pluginName = this.satupPlugins.pop()!;
 
       this.log.debug(`Stopping plugin "${pluginName}"...`);
-      await this.plugins.get(pluginName)!.stop();
+
+      const result = (await Promise.race([
+        this.plugins.get(pluginName)!.stop(),
+        new Promise((resolve) => setTimeout(() => resolve({ delayed: true }), 30 * Sec)),
+      ])) as { delayed?: boolean };
```

Comment:
This pattern might cause some unexpected memory issues in node. We need to clear the `timeout` after the promise has been resolved to avoid build up of zombie timeouts


```
export const PROMISE_TIMEOUT = {};

export function raceTo(promise, ms) {
  let timeout;

  return Promise.race([
    promise,
    new Promise((resolve) => {
      timeout = setTimeout(() => resolve(PROMISE_TIMEOUT), ms);
    }),
  ]).then((res) => {;
    if (timeout) clearTimeout(timeout);

    return res;
  }).catch((err) => {
    if (timeout) clearTimeout(timeout);

    return Promise.reject(err);
  });
}

// ...

if (result === PROMISE_TIMEOUT) {
  this.log.warn(...);
}
```

Reply:
> This pattern might cause some unexpected memory issues in node.

Is there an example of a study of how bad it affects memory consumption? This code is short-living: The Core runs the function ~100 times as a part of the shutdown process. I'm okay to add the proposal, but it doesn't look like a critical functionality. Does it?

Reply:
> but might be unnecessary micro-optimisation.

It is more about a good coding practice rather than an optimization.

You're resolving the promise yet there is a trailing timeout that wants to keep the lexical scope of the wrapping function for no reason. As for the memory effect for this, in addition to preventing the function context to be marked for garbage collection in time, you are adding unncessary cycles to the 'message queue' before triggering the clearing of those timeouts in the "event loop".

In another other words; If all stop functions are resolved immediately and nodejs is ready to exit. You will be holding node from closing the process for an additional 30 seconds until the timeouts are unreffed and cleared from the message queue.


> Is there an example of a study of how bad it affects memory consumption?

I dont have an article handy but this is how nodejs works. You can dig into how the timer works here however if you're interested:

- javascript part: https://github.com/nodejs/node/blob/c0e66e3e9d74d514e3ef8a06a652f45458d98d20/lib/internal/timers.js#L5

-c++ part: https://github.com/nodejs/node/blob/606df7c4e79324b9725bfcfe019a8b75bfa04c3f/deps/v8/src/heap/memory-reducer.cc#L37

Reply:
A slight memory leak while exiting is probably fine, but it would be worth testing that we don't increase the exit delay unnecessarily.

Reply:
> A slight memory leak while exiting is probably fine

I'm slightly confused because this is changing a couple of lines in the implementation. Anyways you can test that it does delay the shutdown by calling this:

```
// node test.js

let timeout1;
let timeout2;

const result = Promise.race([
  new Promise((resolve) => {
    timeout1 = setTimeout(() => resolve('ok'), 100);
  }),
  new Promise((resolve) => {
    timeout2 = setTimeout(() => resolve('ok'), 30000);
  }),
]).then(() => {
  clearTimeout(timeout1), clearTimeout(timeout2);
})
```

If you comment out the `clearTimeout` line you'll get a delay of 30 seconds before the node process exits.

Reply:
> You will be holding node from closing the process for an additional 30 seconds until the timeouts are unreffed and cleared from the message queue.

That's a fair point - I'm going to refactor the function. 
Although, it has nothing to do with performance.

Reply:
You can probably just [`unref()` the timeout objects](https://nodejs.org/dist/latest-v14.x/docs/api/timers.html#timers_timeout_unref).  That's what we used to do ALL THE TIME (in a previous job).  Especially for shutdown logic.  That will keep node from "staying around" until the timeouts "complete".  I believe clearing the timeout will similarly "unref" it, but you can do the unref early, right after the timeout is created.

I don't think I'd worry so much about other resources (lexically accessed) still being held onto.  Kibana will be coming down, so there's probably not much help in marking objects as inaccessible for the GC to reuse.  Hopefully the GC is not allocating (much) more object space during this phase anyway!

As an optimization, it would be good to parallelize any not-related plugins so you can overlap some of these timeouts.

Reply:
That's an option. It's not consistent with `setup` and `start` phases, though: lifecycles are executed in the order defined by the dependency graph. In theory, they can even expose `stop` contracts.